Thanks for the detailed comments Jake, I made the changes recommended by you in both the test as well as the lax_numpy file.

However the tests on x64 are still failing (traceback below).

I've made the edits to the function to look like this:

```
@_wraps(np.sort_complex)
def sort_complex(a):
  a = lax.sort(a, dimension=0)
  return lax.convert_element_type(a, result_type(a, complex_))
```

```
(jax) vaibhavs10@SkyNet:~/jax$ python tests/lax_numpy_test.py --num_generated_cases=50 --jax_enable_x64 --test_targets=testSortComplex
/home/vaibhavs10/jax/jax/lib/xla_bridge.py:130: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
Running tests under Python 3.7.7: /home/vaibhavs10/anaconda3/envs/jax/bin/python
[ RUN      ] LaxBackedNumpyTests.testSortComplex_bfloat16[12]_axis=None
/home/vaibhavs10/jax/jax/lib/xla_bridge.py:130: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
[       OK ] LaxBackedNumpyTests.testSortComplex_bfloat16[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_bfloat16[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_bfloat16[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_bfloat16[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_bfloat16[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_bool[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_bool[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_bool[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_bool[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_bool[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_bool[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_complex128[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_complex128[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_complex128[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_complex128[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_complex128[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_complex128[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_complex64[12]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_complex64[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_complex64[1]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_complex64[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_complex64[6]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_complex64[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float16[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float16[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float16[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float16[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float16[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float16[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float32[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float32[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float32[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float32[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float32[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float32[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float64[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float64[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float64[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float64[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_float64[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_float64[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int16[12]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_int16[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int16[1]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_int16[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int16[6]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_int16[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int32[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_int32[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int32[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_int32[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int32[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_int32[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int64[12]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_int64[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int64[1]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_int64[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int64[6]_axis=None
[       OK ] LaxBackedNumpyTests.testSortComplex_int64[6]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int8[12]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_int8[12]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int8[1]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_int8[1]_axis=None
[ RUN      ] LaxBackedNumpyTests.testSortComplex_int8[6]_axis=None
[  FAILED  ] LaxBackedNumpyTests.testSortComplex_int8[6]_axis=None
======================================================================
FAIL: testSortComplex_complex64[12]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_complex64[12]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_complex64[12]_axis=None(shape=(12,), dtype=<class 'numpy.complex64'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_complex64[1]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_complex64[1]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_complex64[1]_axis=None(shape=(1,), dtype=<class 'numpy.complex64'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_complex64[6]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_complex64[6]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_complex64[6]_axis=None(shape=(6,), dtype=<class 'numpy.complex64'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_int16[12]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int16[12]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int16[12]_axis=None(shape=(12,), dtype=<class 'numpy.int16'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_int16[1]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int16[1]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int16[1]_axis=None(shape=(1,), dtype=<class 'numpy.int16'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_int16[6]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int16[6]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int16[6]_axis=None(shape=(6,), dtype=<class 'numpy.int16'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_int8[12]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int8[12]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int8[12]_axis=None(shape=(12,), dtype=<class 'numpy.int8'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_int8[1]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int8[1]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int8[1]_axis=None(shape=(1,), dtype=<class 'numpy.int8'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

======================================================================
FAIL: testSortComplex_int8[6]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int8[6]_axis=None (__main__.LaxBackedNumpyTests)
testSortComplex_int8[6]_axis=None(shape=(6,), dtype=<class 'numpy.int8'>, axis=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vaibhavs10/anaconda3/envs/jax/lib/python3.7/site-packages/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "tests/lax_numpy_test.py", line 2665, in testSortComplex
    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 881, in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 810, in assertAllClose
    self.assertDtypesMatch(x, y, canonicalize_dtypes=canonicalize_dtypes)
  File "/home/vaibhavs10/jax/jax/test_util.py", line 790, in assertDtypesMatch
    self.assertEqual(_dtype(x), _dtype(y))
AssertionError: dtype('complex128') != dtype('complex64')

----------------------------------------------------------------------
Ran 33 tests in 1.651s

FAILED (failures=9)
```

Should I write a custom logic for this, it always seems to be an issue with Complex128.
Do note that in my test cases for dtypes, I'm currently passing all dtypes (maybe I should only pass the complex ones?)

--- a/jax/numpy/lax_numpy.py
+++ b/jax/numpy/lax_numpy.py
@@ -3240,7 +3240,7 @@ def sort(a, axis=-1, kind='quicksort', order=None):
 @_wraps(np.sort_complex)
 def sort_complex(a):
   a = lax.sort(a, dimension=0)
-  return lax.convert_element_type(a, result_type(a, complex64))
+  return lax.convert_element_type(a, result_type(a, complex_))
 
 @_wraps(np.lexsort)
 def lexsort(keys, axis=-1):

--- a/tests/lax_numpy_test.py
+++ b/tests/lax_numpy_test.py
@@ -2652,7 +2652,7 @@ class LaxBackedNumpyTests(jtu.JaxTestCase):
       {"testcase_name": "_{}_axis={}".format(
           jtu.format_shape_dtype_string(shape, dtype), axis),
         "shape": shape, "dtype": dtype, "axis": axis}
-      for dtype in all_dtypes
+      for dtype in python_scalar_dtypes
       for shape in one_dim_array_shapes
       for axis in [None]))
   def testSortComplex(self, dtype, shape, axis):
@@ -2662,13 +2662,8 @@ class LaxBackedNumpyTests(jtu.JaxTestCase):
       self.skipTest("complex sort not supported on TPU")
     rng = jtu.rand_some_equal(self.rng())
     args_maker = lambda: [rng(shape, dtype)]
-    jnp_fun = jnp.sort_complex
-    np_fun = np.sort_complex
-    if axis is not None:
-      jnp_fun = partial(jnp_fun, axis=axis)
-      np_fun = partial(np_fun, axis=axis)
-    self._CheckAgainstNumpy(jnp_fun, np_fun, args_maker)
-    self._CompileAndCheck(jnp_fun, args_maker)
+    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
+    self._CompileAndCheck(jnp.sort_complex, args_maker)
 
   @parameterized.named_parameters(jtu.cases_from_list(
       {"testcase_name": "_{}_input_type={}_axis={}".format(

--- a/tests/lax_numpy_test.py
+++ b/tests/lax_numpy_test.py
@@ -2652,7 +2652,7 @@ class LaxBackedNumpyTests(jtu.JaxTestCase):
       {"testcase_name": "_{}_axis={}".format(
           jtu.format_shape_dtype_string(shape, dtype), axis),
         "shape": shape, "dtype": dtype, "axis": axis}
-      for dtype in python_scalar_dtypes
+      for dtype in all_dtypes
       for shape in one_dim_array_shapes
       for axis in [None]))
   def testSortComplex(self, dtype, shape, axis):
@@ -2662,7 +2662,7 @@ class LaxBackedNumpyTests(jtu.JaxTestCase):
       self.skipTest("complex sort not supported on TPU")
     rng = jtu.rand_some_equal(self.rng())
     args_maker = lambda: [rng(shape, dtype)]
-    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker)
+    self._CheckAgainstNumpy(jnp.sort_complex, np.sort_complex, args_maker, check_dtypes=False)
     self._CompileAndCheck(jnp.sort_complex, args_maker)
 
   @parameterized.named_parameters(jtu.cases_from_list(

